File: 2/2_cuatrimestre/PDOO/practicas/practica 4/deepSpaceRuby/Dice.rb
require_relative 'GameCharacter'

module Deepspace

    class Dice

        def initialize
            @NHANGARSPROB = 0.25
            @NSHIELDSPROB = 0.25
            @NWEAPONSPROB = 0.33
            @FIRSTSHOTPROB = 0.5
            @EXTRAEFFICIENCYPROB = 0.8
            @generator = Random.new
        end

        def initWithNHangars
            if (@generator.rand <= @NHANGARSPROB)
                0
            else
                1
            end
        end

        def initWithNWeapons
            random = @generator.rand
            if (random <= @NWEAPONSPROB)
                1
            elsif (random <= 2*@NWEAPONSPROB)
                2
            else
                3
            end         
        end

        def initWithNShields
            if (@generator.rand <= @NSHIELDSPROB)
                0
            else
                1
            end
        end

        def whoStarts (nPlayers)
            @generator.rand(nPlayers)
        end

        def firstShot
            if (@generator.rand <= @FIRSTSHOTPROB)
                GameCharacter::SPACESTATION
            else
                GameCharacter::ENEMYSTARSHIP
            end
        end
        
        def spaceStationMoves (speed)
            @generator.rand <= speed            
        end

        def extraEfficiency
            @generator.rand <= @EXTRAEFFICIENCYPROB
        end
    end


end

File: 2/2_cuatrimestre/PDOO/practicas/practica 4/deepSpace_4/src/deepspace/Dice.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

import java.util.Random;

/**
 *
 * @author manbolq
 */
public class Dice {
    private final float NHANGARSPROB;
    private final float NSHIELDSPROB;
    private final float NWEAPONSPROB;
    private final float FIRSTSHOTPROB;
    private final float EXTRAEFFICIENCYPROB;
    
    private Random generator;
    
    
    /**************************************************************************/
    /*************************    CONSTRUCTORES   *****************************/
    /**************************************************************************/
    
    Dice(){
        NHANGARSPROB=0.25f;
        NSHIELDSPROB=0.25f;
        NWEAPONSPROB=0.33f;
        FIRSTSHOTPROB=0.5f;
        EXTRAEFFICIENCYPROB=0.8f;
        generator = new Random();
    }
    
    
    /**************************************************************************/
    /****************************    METODOS   ********************************/
    /**************************************************************************/
    
    int initWithNHangars(){
        float random = generator.nextFloat();
        if (random <= NHANGARSPROB)
            return 0;
        else
            return 1;
    }
    
    int initWithNWeapons(){
        float random = generator.nextFloat();
        if (random <= NWEAPONSPROB)
            return 1;
        else if (random <= 2*NWEAPONSPROB)
            return 3;
        else
            return 2;
    }
    
    
    int initWithNShields(){
        float random = generator.nextFloat();
        
        if (random <= NSHIELDSPROB)
            return 0;
        else
            return 1;
    }
    
    
    int whoStarts(int nPlayers){
        return generator.nextInt(nPlayers);
    }
    
    
    GameCharacter firstShot(){
        float random = generator.nextFloat();
        if (random <= FIRSTSHOTPROB)
            return GameCharacter.SPACESTATION;
        else
            return GameCharacter.ENEMYSTARSHIP;            
    }
    
    
    boolean spaceStationMoves(float speed){
        float random = generator.nextFloat();
        return (random <= speed);
    }
    
    boolean extraEfficiency(){
        return (generator.nextFloat() <= EXTRAEFFICIENCYPROB);
    }
}
